### PR TITLE
Base64 encode the "code" query parameter.

### PR DESCRIPTION
**Pull request by whipsch**
This changes the "code" query parameter to "enc", encoded as URL-safe base64.
The "code" query parameter for unencoded code is still present, allowing old playpen links to remain functional.

Addons such as NoScript would often flag playpen URLs as potential XSS attacks, due to liberal usage of sequences such as \<T\> in Rust code. Encoding the query parameter prevents this from happening.

Unfortunately, this results in somewhat longer URLs.  However, even with the encoding, the 5000 character limit is still quite permissive for typical playpen usage.

### CODE REVIEW

File: static/web.js [hunk 1]
```diff
@@ -69,11 +69,23 @@ function format(result, session, version) {
     });
 }
 
+// URL safe alphabet, no padding: https://tools.ietf.org/html/rfc4648#section-5
+// encodeURIComponent is due to https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64.btoa#Unicode_Strings
+function b64_encode(str) {
+    return btoa(encodeURIComponent(str)).replace(/\+/g, '-').replace(/\//g, '_').replace(/\=+$/, '');
```
Comment (alexcrichton) [inline static/web.js]: Are the two layers of encoding necessary? Could `encodeURIComponent` be eschewed in favor of just base64 encoding?

**Comment by whipsch** [comment]
The reason for the extra layer of encoding is outlined on this page: https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64.btoa#Unicode_Strings

... however I left out the extra usage of unescape, which turns out to be quite significant.  It seems the code on that page intentionally clobbers any unicode characters via unescape() after encoding with encodeURIComponent(), which encodes them correctly.  The reverse is then performed when decoding (re-encode with escape(), then decode it properly via decodeURIComponent() to restore any unicode characters).  

The code I submitted should in fact use both functions as currently it will unnecessarily bloat the size of the encoded data.

However, I'm not sure if being totally unicode-safe with the catch of using wholly deprecated functions is worth it.  Regardless, the single extra encodeURIComponent() should not be there without the accompanying unescape().  Perhaps it would be better to include non-internal base64 routines that properly handle unicode?

**Comment by alexcrichton** [comment]
Could we perhaps take a more conservative route where only the `<` and `>` characters are encoded? If that's really the only problem with various extensions, then perhaps that's all that needs to be taken care of?

**Comment by whipsch** [comment]
Forgot about this PR for a while, but ultimately I've reached the conclusion that this isn't really worth it.

The best solution for encoding the flagged characters without bloating the URL that I've come up with is to replace them with the antiquated ASCII control characters (like BEL, ACK, ENQ, etc).  While it would work, I don't really like that solution very much and it just smells very hacky.

NoScript users can add the following exception pattern instead (NoScript Options -> Advanced -> XSS):
```
^https://play\.rust-lang\.org/\?code=
```

**Pull request closed without merging**